llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -9415,6 +9415,13 @@ bool RISCVTargetLowering::isFMAFasterThanFMulAndFAdd(const MachineFunction &MF,
   return false;
 }
 
+unsigned RISCVTargetLowering::getMinCmpXchgSizeInBits(const Instruction *I,
+                                                      const Value *Ptr) const {
+  if (I->getModule()->getDataLayout().isFatPointer(Ptr->getType()))
+    return 8; // All sizes are supported via capabilities
+  return TargetLowering::getMinCmpXchgSizeInBits(I, Ptr);
```

Comment:
I think that's because of this code:
```
TargetLowering::AtomicExpansionKind
RISCVTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
  // atomicrmw {fadd,fsub} must be expanded to use compare-exchange, as floating
  // point operations can't be used in an lr/sc sequence without breaking the
  // forward-progress guarantee.
  if (AI->isFloatingPointOperation())
    return AtomicExpansionKind::CmpXChg;

  unsigned Size = AI->getType()->getPrimitiveSizeInBits();
  if ((Size == 8 || Size == 16) &&
      !RISCVABI::isCheriPureCapABI(Subtarget.getTargetABI()))
    return AtomicExpansionKind::MaskedIntrinsic;
  return AtomicExpansionKind::None;
}
```

Reply:
Not sure if I understand the question, but i8/i16 ops are expanded in IR: https://godbolt.org/z/Y9bKxeb88